internal/daemon/api_logs.go

```diff
@@ -0,0 +1,318 @@
+		go func(service string, it servicelog.Iterator) {
+			defer wg.Done()
+			reader := bufio.NewReaderSize(it, bufioReaderSize)
+			fw := &followWriter{w, writeMutex}
+			encoder := json.NewEncoder(fw)
+			encoder.SetEscapeHTML(false)
+			err := iterateLogs(it, reader, done, time.Time{}, func(log logEntry) error {
+				return encoder.Encode(log)
+			})
+			if err != nil {
+				fmt.Fprintf(w, "\nerror writing logs: %v", err)
+				return
+			}
+		}(service, it)
+	}
+
+	// Don't return till client connection is closed and all iterateLogs have
+	// finished. We can't just wait for the done channel here, because the
+	// client connection will close and iterateLogs may still be calling
+	// it.Next() when the caller calls it.Close(), causing a data race.
+	wg.Wait()
+}
+
+// Each log is written as a JSON object followed by a newline (JSON Lines):
+//
+// {"time":"2021-04-23T01:28:52.660Z","service":"redis","message":"redis started up"}
+// {"time":"2021-04-23T01:28:52.798Z","service":"thing","message":"did something"}
+type logEntry struct {
+	Time    time.Time `json:"time"`
+	Service string    `json:"service"`
+	Message string    `json:"message"`
+}
+
+type writeLogFunc func(log logEntry) error
+
+// Iterate through all logs in the given iterator, calling writeLog for each
+// until the iterator is exhausted.
+func iterateLogs(it servicelog.Iterator, br *bufio.Reader, cancel <-chan struct{}, now time.Time, writeLog writeLogFunc) error {
+	var prevTime time.Time
+	var prevService string
+	for it.Next(cancel) {
+		br.Reset(it)
+		for {
+			line, err := br.ReadSlice('\n')
+			if err != nil && !errors.Is(err, io.EOF) {
+				return err
+			}
+			if len(line) == 0 {
+				// No bytes read, we're done with this iteration
+				break
+			}
+
+			log, ok := parseLog(line)
+			if ok {
+				// Output normal log
+				err := writeLog(log)
+				if err != nil {
+					return err
+				}
+				if !now.IsZero() && log.Time.After(now) {
+					// Stop if we're past current time
+					return nil
+				}
+				prevTime = log.Time
+				prevService = log.Service
+			} else if !prevTime.IsZero() {
+				// Output partial log with previous timestamp and service name
+				err := writeLog(logEntry{prevTime, prevService, string(line)})
+				if err != nil {
+					return err
+				}
+			}
+
+			if errors.Is(err, io.EOF) {
+				// EOF reached, break after processing line
+				break
+			}
+		}
+	}
+	return nil
+}
+
+// Parse a log entry of the form "2021-05-20T15:39:12.345Z [service] log message"
+func parseLog(line []byte) (logEntry, bool) {
+	fields := bytes.SplitN(line, []byte(" "), 3)
+	if len(fields) != 3 {
+		return logEntry{}, false
+	}
+	// .999 allows any number of fractional seconds (including none at all)
+	timestamp, err := time.Parse("2006-01-02T15:04:05.999Z07:00", string(fields[0]))
+	if err != nil {
+		return logEntry{}, false
+	}
+	if len(fields[1]) < 3 {
+		return logEntry{}, false
+	}
+	service := string(fields[1][1 : len(fields[1])-1]) // Trim [ and ] from "[service]"
```

Comment:
We need to check and complain loudly if it does not conform.

Reply:
Fixed. It returns an error if the `[` or `]` are missing now.